### PR TITLE
Observing manipulations

### PR DESCRIPTION
**Pull request by ldeichmann**
Adds a new kind of precondition which receives changes to the Mdib of the Device under Test and can react accordingly.

# Checklist

The following aspects have been respected by the author of this pull request, confirmed by both pull request assignee **and** reviewer:

* Adherence to coding conventions
  * [x] Pull Request Assignee
  * [x] Reviewer
* Adherence to javadoc conventions
  * [x] Pull Request Assignee
  * [x] Reviewer
* Changelog update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer
* README update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer
* config update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer
* SDCcc executable ran against a test device (if necessary)
  * [x] Pull Request Assignee
  * [x] Reviewer

### CODE REVIEW

File: sdccc/src/main/java/com/draeger/medical/sdccc/tests/biceps/invariant/InvariantBicepsNormativeAnnexTest.java [hunk 1]
```diff
@@ -49,6 +51,7 @@ void setUp() {
     @TestIdentifier(EnabledTestConfig.BICEPS_R5006)
     @TestDescription(
             "Verifies that for all incoming messages, Localized text with the @Ref attribute always has a version")
+    @RequirePrecondition(observingPreconditions = {DummyObservingPrecondition.Companion.class})
```
Comment (ldeichmann) [inline sdccc/src/main/java/com/draeger/medical/sdccc/tests/biceps/invariant/InvariantBicepsNormativeAnnexTest.java:54]: Remove

File: sdccc/src/main/java/com/draeger/medical/sdccc/manipulation/precondition/impl/DummyObservingPrecondition.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import com.draeger.medical.sdccc.manipulation.precondition.Observing
+import com.draeger.medical.sdccc.manipulation.precondition.ObservingPreconditionFactory
+import com.draeger.medical.sdccc.sdcri.testclient.MdibChange
+import com.draeger.medical.sdccc.sdcri.testclient.TestClient
+import com.google.inject.Injector
+
+/**
+ * An example of an observing precondition.
+ */
+class DummyObservingPrecondition(injector: Injector) : BufferedObservingPrecondition(
```
Comment (ldeichmann) [inline sdccc/src/main/java/com/draeger/medical/sdccc/manipulation/precondition/impl/DummyObservingPrecondition.kt:20]: Remove

**Review by pstaeck — CHANGES_REQUESTED** [review]

File: sdccc/src/main/java/com/draeger/medical/sdccc/tests/annotations/RequirePrecondition.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    val simplePreconditions: Array<KClass<out SimplePrecondition>> = [],
+    /**
+     * Manipulations to trigger.
+     *
+     * @return array of manipulation classes to trigger for the test
+     */
+    val manipulationPreconditions: Array<KClass<out ManipulationPrecondition>> = [],
+    /**
+     * Manipulations to trigger.
```
Comment (belagertem) [inline sdccc/src/main/java/com/draeger/medical/sdccc/tests/annotations/RequirePrecondition.kt:43]: Please adapt KDoc.

File: sdccc/src/main/java/com/draeger/medical/sdccc/manipulation/precondition/PreconditionRegistry.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+     * Duplicate preconditions will be ignored.
+     *
+     * @param precondition precondition to run
+     */
+    fun registerManipulationPrecondition(precondition: Class<out ManipulationPrecondition?>) {
+        registerPreconditionInternal(precondition)
+    }
+
+    /**
+     * Registers a manipulation precondition for running and observing before disconnecting from the DUT.
```
Comment (belagertem) [inline sdccc/src/main/java/com/draeger/medical/sdccc/manipulation/precondition/PreconditionRegistry.kt:77]:
```suggestion
     * Registers an observing precondition for running and observing before disconnecting from the DUT.
```

**Review by belagertem — CHANGES_REQUESTED** [review]

File: sdccc/src/main/java/com/draeger/medical/sdccc/sdcri/testclient/TestClientMdibObserver.kt [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    /**
+     * Represents a change to the alert states of the device.
+     *
+     * @property mdibVersion      the MDIB version that the change is occurring in.
+     * @property states the modification where the key is the source MDS and
+     *                  the value is a list of all states for that MDS.
+     */
+    data class Alert(
+        override val mdibVersion: MdibVersion,
+        val states: Map<String, List<AbstractAlertState>>
```
[the remark this answers is not in the document]
Comment (ldeichmann) [reply sdccc/src/main/java/com/draeger/medical/sdccc/sdcri/testclient/TestClientMdibObserver.kt:115]: I think all names are kinda dumb here, `statesMap` e.g. makes it sound like there is an alternative way to access states, but there isn't. I decided to keep the naming from the sdc-ri message we get them from

**Review by pstaeck — APPROVED** [review]

**Review by jannast — APPROVED** [review]

**Review by belagertem — APPROVED** [review]

**Pull request merged**